### PR TITLE
DaemonSet: Respect ControllerRef

### PR DESCRIPTION
**Pull request by enisoc**
**What this PR does / why we need it**:

This is part of the completion of the [ControllerRef](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md) proposal. It brings DaemonSet into full compliance with ControllerRef. See the individual commit messages for details.

**Which issue this PR fixes**:

This ensures that DaemonSet does not fight with other controllers over control of Pods.

**Special notes for your reviewer**:

**Release note**:

```release-note
DaemonSet now respects ControllerRef to avoid fighting over Pods.
```
cc @erictune @kubernetes/sig-apps-pr-reviews

### CODE REVIEW

**Comment by kow3ns** [comment]
Reviewed 33 of 33 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/controller/daemon/daemoncontroller.go, line 73 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42173#-Ke0F7a-G0UpKNBMEqar:-Ke0F7a-G0UpKNBMEqas:bixxmvg) ([raw file](https://github.com/kubernetes/kubernetes/blob/f2cf9d9986d94ebb5479476a5f6078646abc610d/pkg/controller/daemon/daemoncontroller.go#L73)):*
> ```Go
> 
> // ControllerKind contains the schema.GroupVersionKind for this controller type.
> var ControllerKind = extensions.SchemeGroupVersion.WithKind("DaemonSet")
> ```

You sure you want this to be package public?

---

*[pkg/controller/daemon/daemoncontroller.go, line 103 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42173#-Ke0FoZCJZcQBs_gMqNo:-Ke0FoZCJZcQBs_gMqNp:brkrjt6) ([raw file](https://github.com/kubernetes/kubernetes/blob/7f626cf8365a96d3fbdbe9510386321f7b5d0223/pkg/controller/daemon/daemoncontroller.go#L103)):*
> ```Go
> 	nodeStoreSynced cache.InformerSynced
> 
> 	lookupCache *controller.MatchingCache
> ```

Are we removing the cache because shared

---

*[pkg/controller/replicaset/replica_set.go, line 62 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42173#-Ke0It1NseI33OEGA6KH:-Ke0It1O_9UHebvw-ZeV:b-kk6lw0) ([raw file](https://github.com/kubernetes/kubernetes/blob/f2cf9d9986d94ebb5479476a5f6078646abc610d/pkg/controller/replicaset/replica_set.go#L62)):*
> ```Go
> 
> // ControllerKind contains the schema.GroupVersionKind for this controller type.
> var ControllerKind = v1beta1.SchemeGroupVersion.WithKind("ReplicaSet")
> ```

You sure you want to export this?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42173)*

**Comment by kow3ns** [comment]
*[pkg/controller/daemon/daemoncontroller.go, line 103 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42173#-Ke0FoZCJZcQBs_gMqNo:-Ke0JmtOWtPY-ejvAWLK:binvtqr) ([raw file](https://github.com/kubernetes/kubernetes/blob/7f626cf8365a96d3fbdbe9510386321f7b5d0223/pkg/controller/daemon/daemoncontroller.go#L103)):*
<details><summary><i>Previously, kow3ns (Kenneth Owens) wrote…</i></summary><blockquote>

Are we removing the cache because shared
</blockquote></details>

informers make it no longer necessary?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42173)*

**Comment by kow3ns** [comment]
Other than the comments lgtm.  Are you removing the cache entirely because the shared informers mitigate its usefulness.

---







*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42173#-:-Ke0Oh39JlyTV6qv80IT:bcgyuz4)*

File: pkg/controller/controller_utils.go
Comment (caesarxuchao) [inline pkg/controller/controller_utils.go]: Could you also verify that controllerRef.BlockOwnerDeletion is true, and set it to true in all controllee's creation/adoption? See the [comments](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L259-L266) here.

File: pkg/controller/daemon/daemoncontroller.go
Comment (caesarxuchao) [inline pkg/controller/daemon/daemoncontroller.go]: Could you set the BlockOwnerDeletion like what we did for replica set controller [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replicaset/replica_set.go#L490-L491)? Please update all for all the controllers.

Comment (enisoc) [inline pkg/controller/daemon/daemoncontroller.go]: Done.

File: pkg/controller/controller_utils.go
Comment (enisoc) [inline pkg/controller/controller_utils.go]: I will do this in a separate PR since the change will apply to all controllers.

Comment (enisoc) [inline pkg/controller/controller_utils.go]: After looking more closely, it makes more sense to fix it here. I just needed to move the BlockOwnerDeletion check up into validateControllerRef().

File: pkg/controller/controller_utils.go
Comment (kargakis) [inline pkg/controller/controller_utils.go]: Is this validation running in the api server or only in controllers?

Comment (caesarxuchao) [inline pkg/controller/controller_utils.go]: Only in controllers.

**Comment by kargakis** [comment]
Should we have it in the api server too for TPRs?

File: pkg/controller/controller_utils.go
Comment (caesarxuchao) [inline pkg/controller/controller_utils.go]: This validation is only required for controllers. API server doesn't know if a request is sent by a controller thus cannot do the validation.

If a TPR creates dependents, then its custom controller could do a similar check.

**Comment by enisoc** [comment]
@k8s-bot gce etcd3 e2e test this
@k8s-bot gci gce e2e test this

File: cmd/kube-controller-manager/app/options/options.go
Comment (janetkuo) [inline cmd/kube-controller-manager/app/options/options.go]: We should call `fs.MarkDeprecated`, and put a TODO for when to remove this flag

Comment (enisoc) [inline cmd/kube-controller-manager/app/options/options.go]: Thanks I fixed it here and added it to my list of fixes for RC/RS that I will make in a separate PR.

File: pkg/controller/daemon/daemoncontroller.go
Comment (janetkuo) [inline pkg/controller/daemon/daemoncontroller.go]: We should probably add a test to catch this?

File: pkg/controller/controller_utils.go
Comment (enisoc) [inline pkg/controller/controller_utils.go]: I'll make a note to address that in a future PR. It will affect multiple controllers and may require updates to their unit tests, since not all tests bother to fill those in. In this PR, I am just moving this validation code to a common function so I can use it in two places without duplicating.

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (janetkuo) [inline pkg/controller/daemon/daemoncontroller_test.go]: also check the items in the queue

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (janetkuo) [inline pkg/controller/daemon/daemoncontroller_test.go]: ditto

File: pkg/controller/daemon/daemoncontroller.go
Comment (enisoc) [inline pkg/controller/daemon/daemoncontroller.go]: In another ControllerRef PR, the reviewers asked me to remove this because it's too noisy for a watch event handler. I applied the change across all my ControllerRef PRs. The sync loop is a bit different because it's specific to DaemonSet, whereas the watch even handlers get called for many events that they end up not caring about.

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (janetkuo) [inline pkg/controller/daemon/daemoncontroller_test.go]: `...WithMatchingLabels`

File: pkg/controller/daemon/daemoncontroller.go
Comment (enisoc) [inline pkg/controller/daemon/daemoncontroller.go]: I think that's a separate issue since I'm not changing the semantics in this PR. Note that the previous line `daemonPod := &(*daemonPods[i])` was not actually doing a copy ([example](https://play.golang.org/p/wtLp9uxU8O)).

Can we discuss this after landing ControllerRef? We should also think about the general pattern to use across controllers.

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (janetkuo) [inline pkg/controller/daemon/daemoncontroller_test.go]: Isn't this adopt rather than release?

File: pkg/controller/daemon/daemoncontroller.go
Comment (enisoc) [inline pkg/controller/daemon/daemoncontroller.go]: In other controllers, we expect changes we make to be observed only in the next sync, and we re-queue when appropriate. Does DaemonSet not work that way?

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (enisoc) [inline pkg/controller/daemon/daemoncontroller_test.go]: In general, it's good practice to make unit test expected values hard-coded whenever possible. This helps to avoid problems being hidden due to tests that are actually tautologies: "check that A == B but actually both come from C, so the test passes even if C is wrong."

In particular, having hard-coded values in this test saved me in the CronJob PR when I rebased and someone had renamed a package on import. That actually made `controllerKind` use the wrong `SchemeGroupVersion`, but this test caught it.

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (enisoc) [inline pkg/controller/daemon/daemoncontroller_test.go]: The name might be confusing. It is testing the case that a pod *was* released and we are observing the update event. You're right that we are expecting the reaction to be "queue for potential adoption".

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (enisoc) [inline pkg/controller/daemon/daemoncontroller_test.go]: Fixed in upcoming push to the PR.

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (enisoc) [inline pkg/controller/daemon/daemoncontroller_test.go]: Fixed in upcoming push to the PR.

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (enisoc) [inline pkg/controller/daemon/daemoncontroller_test.go]: Fixed in upcoming push to the PR.

**Comment by enisoc** [comment]
@k8s-bot gci gce e2e test this
@k8s-bot kops aws e2e test this
@k8s-bot cvm gce e2e test this
@k8s-bot non-cri e2e test this
@k8s-bot gce etcd3 e2e test this

File: pkg/controller/daemon/daemoncontroller.go
Comment (janetkuo) [inline pkg/controller/daemon/daemoncontroller.go]: sounds good

File: pkg/controller/controller_utils.go
Comment (janetkuo) [inline pkg/controller/controller_utils.go]: sounds good

File: pkg/controller/daemon/daemoncontroller.go
Comment (janetkuo) [inline pkg/controller/daemon/daemoncontroller.go]: Yes we can address this later

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (janetkuo) [inline pkg/controller/daemon/daemoncontroller_test.go]: Thanks for the explanation

File: pkg/controller/daemon/daemoncontroller.go
Comment (janetkuo) [inline pkg/controller/daemon/daemoncontroller.go]: No, DaemonSet controller didn't work that way. In the rollingUpdate case, we may create/kill some pods after `manage`, and then do the rolling update based on `nodeToDaemonPods`. If we use the cached value after we changed those pods in `manage`, rolling update may create or kill more pods than needed. It might be fine for Deployments, but it's safer not to do this in DaemonSet controller (since every node expects one and exactly one pod.)

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (janetkuo) [inline pkg/controller/daemon/daemoncontroller_test.go]: The new name looks good

File: pkg/controller/daemon/daemoncontroller.go
Comment (kow3ns) [inline pkg/controller/daemon/daemoncontroller.go]: StatefulSet is now deep copying everything FWIW.

**Comment by caesarxuchao** [comment]
/approve

Approve the changes to client.

File: pkg/controller/daemon/daemoncontroller.go
Comment (enisoc) [inline pkg/controller/daemon/daemoncontroller.go]: OK I switched back to relisting before each phase, and added a test. PTAL.

**Comment by janetkuo** [comment]
LGTM. Maybe you want to squash commits first?

**Comment by enisoc** [comment]
I usually retain separate commits that represent discrete "layers" of a PR as described here: https://github.com/kubernetes/community/blob/master/contributors/devel/faster_reviews.md#8-fix-feedback-in-a-new-commit

File: cmd/kube-controller-manager/app/options/options.go
Comment (janetkuo) [inline cmd/kube-controller-manager/app/options/options.go]: Copying from my https://github.com/kubernetes/kubernetes/pull/42659/files#r104817545:

> `MarkDeprecated` prints deprecation warnings for you, but we still need these flags.

**Comment by janetkuo** [comment]
One last comment and LGTM otherwise

File: cmd/kube-controller-manager/app/options/options.go
Comment (enisoc) [inline cmd/kube-controller-manager/app/options/options.go]: Done.

**Review by janetkuo — APPROVED** [review]

**Comment by lavalamp** [comment]
/approve

(for pkg/registry, componentconfig, and kube-controller-manager changes only)

**Comment by enisoc** [comment]
@k8s-bot bazel test this
@k8s-bot gci gce e2e test this

**Comment by liggitt** [comment]
was there agreement that the controller ref proposal would be implemented post-code-freeze? this proposal seems to involve rewriting all the controllers. these are significant changes dropping extremely late in the release.

**Comment by enisoc** [comment]
@liggitt You're right that I was late, and I make no excuses for that. I was asked to keep working on this because the goal is an important stability improvement with the potential to significantly lessen the pager load for us over the next quarter and reduce confusing breakages for users.

Given that, we classified this as bugfix/stability work, and I tried to follow the procedures for that. Although the devil is always in the details, we think these changes are low risk relative to the known problems they address, because the insertion points are clearly defined and largely orthogonal to the primary business logic of controllers.

In particular, the main insertion points are:

1. Apply ControllerRef to claim ownership of objects you create.
2. When listing objects to manage, filter out anything you don't own.
3. When a watch event is received, route it to the owning controller based on ControllerRef.

Points 1 and 2 were already implemented for RC and RS in v1.5, and I have now copied that pattern to the other controllers. Point 3 fixes a remaining bug that could result in RC/RS stalling and never syncing. The real meat of the controllers is in the actual managing of those objects once they're found, and I haven't touched those parts.

If you think this needs to be discussed with a larger audience, please include me, @janetkuo and @erictune in your communication. Thanks!

**Comment by lavalamp** [comment]
I added do-not-merge to the others in this series until we resolve this.

**Comment by lavalamp** [comment]
FWIW, I think we really need these changes, they are blocking other improvements.

**Comment by liggitt** [comment]
I agree fighting controllers is bad, but these PRs are the definition of "planned bug fixes", which we've explicitly said we weren't doing.

Process aside, I just want a stable release. These rewrites are big enough that I can't tell whether they will help with a stable release or not. We know we have problems with fighting controllers, but these could be trading those for unknown problems with orphaning, etc.

In general, I'm in favor of the changes, but I'm concerned we're losing all our soak time for 1.6. At the very least, the decision to make these changes to complete a proposal should have been done transparently via the exception process.

**Comment by erictune** [comment]
@ethernetdan   there is a difference of opinion on whether this PR was a "just a bug fix" and thus able to go in after the freeze, or whether it changes enough that it is a feature, and thus should not have merged after the freeze (it was only lgmt'ed after the freeze).

**Comment by lavalamp** [comment]
(and #42175, #42176, #42177, #42659 are all in the same boat)

**Comment by liggitt** [comment]
The "this is part of completing a proposal" comment really makes it seem like a feature, albeit one intended to stabilize and improve the controllers. 

Either way, we should decide quickly. If it is the judgement of the relevant owners and release leads that the benefit outweighs the risk, that's great, and we should get what soak time we can. Maybe that discussion already happened, there was just no visibility to it.

**Comment by ethernetdan** [comment]
> @ethernetdan there is a difference of opinion on whether this PR was a "just a bug fix" and thus able to go in after the freeze, or whether it changes enough that it is a feature, and thus should not have merged after the freeze (it was only lgmt'ed after the freeze).

One of the takeaways from this process is that we need to have clearer guidelines on constitutes a bug for code freeze. With the size of these changes it probably should have gone through the exception process and were other similar changes that did not have their exceptions accepted.

> Either way, we should decide quickly. If it is the judgement of the relevant owners and release leads that the benefit outweighs the risk, that's great, and we should get what soak time we can. Maybe that discussion already happened, there was just no visibility to it.

That discussion hasn't occurred yet. My current thinking given where we are in the release process is we move in the direction that least jeopardizes the release:
* If we feel confident in the rest of the PRs, we continue moving forward with them
* If we don't feel confident in either the remaining PRs or the ones that merged, we should revert

Regardless we should try to move quickly here.
Could all relevant reviewers/owners please weigh in:  @enisoc,@kargakis, @kow3ns, @janetkuo, @lukaszo, @caesarxuchao. Let's try to make a decision by EOD if possible.

**Comment by enisoc** [comment]
@ethernetdan Just to clarify, the remaining unmerged PRs are independent of the ones that merged, so a third option is to stop here and neither merge nor revert anything.

As for weighing in, the summary of my opinion in https://github.com/kubernetes/kubernetes/pull/42173#issuecomment-284965630 is that the risk is low because RC/RS were already doing this since v1.5. If we stop with what has merged so far, the net result is that in v1.6 we will roll out that pattern to a few more controllers: SS, DS, Deployment. As I understand it, these are the ones that are causing the most user/ops pain currently.

**Comment by ethernetdan** [comment]
@enisoc thanks for the clarification. In that case I would propose that we keep what's merged and hold off on others until 1.7 - unless they're any objections?

**Comment by kargakis** [comment]
Fine by me

**Comment by liggitt** [comment]
It appears the ones already merged are causing issues with garbage collection: https://github.com/kubernetes/kubernetes/issues/42639

**Comment by enisoc** [comment]
We are discussing the GC issue on https://github.com/kubernetes/kubernetes/issues/42639, but for anyone following here: That flake appears to be due to a known race condition (item 6 in https://github.com/kubernetes/kubernetes/issues/26120) that predates my PRs and already exists in v1.5 for RC/RS.

**Pull request merged**
